src/components/implementation/no_interface/new_boot/boot_deps.h

```diff
@@ -79,65 +89,80 @@ boot_comp_map_memory(struct cobj_header *h, spdid_t spdid, pgtblcap_t pt)
 	return 0;
 }
 
-void
-cos_upcall_fn(upcall_type_t t, void *arg1, void *arg2, void *arg3)
+/* Initialize just the captblcap and pgtblcap, due to hack for upcall_fn addr */
+static void
+boot_compinfo_init(int spdid, captblcap_t *ct, pgtblcap_t *pt, u32_t vaddr)
 {
-	static int first = 1;
-	int i;
+	*ct = cos_captbl_alloc(&boot_info);
+	assert(*ct);
+	*pt = cos_pgtbl_alloc(&boot_info);
+	assert(*pt);
+	new_comp_cap_info[spdid].compinfo = &new_compinfo[spdid];
+	cos_compinfo_init(new_comp_cap_info[spdid].compinfo, *pt, *ct, 0, 
+				  (vaddr_t)vaddr, 4, &boot_info);
+}
+
+static void
+boot_newcomp_create(int spdid, captblcap_t ct, pgtblcap_t pt)
+{
+		compcap_t cc;
+		sinvcap_t sinv;
+		thdcap_t main_thd;
+		int i = 0;
+			
+		cc = cos_comp_alloc(&boot_info, ct, pt, (vaddr_t)new_comp_cap_info[spdid].upcall_entry);
+		assert(cc);	
+		new_comp_cap_info[spdid].compinfo->comp_cap = cc;
 	
-	printc("core %ld: <<cos_upcall_fn thd %d (type %d, CREATE=%d, DESTROY=%d, FAULT=%d)>>\n",
-	       cos_cpuid(), cos_get_thd_id(), t, COS_UPCALL_THD_CREATE, COS_UPCALL_DESTROY, COS_UPCALL_UNHANDLED_FAULT);
+		/* Create sinv capability from Userspace to Booter components */
+		sinv = cos_sinv_alloc(&boot_info, boot_info.comp_cap, (vaddr_t)__inv_test_entry);
+		assert(sinv > 0);
+		
+		/* Copy sinv into new comp's capability table at a known location (BOOT_SINV_CAP) */
+		cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_SINV_CAP, &boot_info, sinv);
 
-	if (first) {
-		first = 0;
-		for(i = 0; i < MAX_NUM_SPDS; i++){
-			schedule[i] = NULL;
-		}
-		__alloc_libc_initilize();
-		constructors_execute();
-		cur_sched = 0;
-	}
+		main_thd = cos_initthd_alloc(&boot_info, cc);
+		assert(main_thd);
 
-	switch (t) {
-	case COS_UPCALL_THD_CREATE:
-	{
-		/* arg1 is the thread init data. 0 means
-		 * bootstrap. */
-		if (arg1 == 0) {
-			cos_init(NULL);
-		}
-	
-		cos_thd_switch(schedule[cur_sched]);
-	
-		return;
-	}
-	default:
-		/* fault! */
-		*(int*)NULL = 0;
-		return;
-	}
-	
-	return;
+		/* Add created component to "scheduling" array */		
+		while (schedule[i] != NULL) i++;
+		schedule[i] = main_thd;
 }
 
-#define CAP_ID_32B_FREE BOOT_CAPTBL_FREE            // goes up
-capid_t capid_32b_free = CAP_ID_32B_FREE;
+static void
```

Comment:
I think this is pretty generic.  Can't this be in booter.c?  I believe it is for the llbooter.